src/Microsoft.Sbom.Parsers.Spdx30SbomParser/Utils/InternalMetadataProviderIdentityExtensions.cs

```diff
@@ -0,0 +1,190 @@
+        }
+
+        // Right now we don't have any better way to version the package. Throw an exception for the user to
+        // provide a package version.
+        throw new ArgumentException($"Unable to generate a package version based on provided parameters. " +
+                                    $"Please provide the package version in the 'PackageVersion' parameter.");
+    }
+
+    public static string GetPackageSupplier(this IInternalMetadataProvider internalMetadataProvider)
+    {
+        if (internalMetadataProvider is null)
+        {
+            throw new ArgumentNullException(nameof(internalMetadataProvider));
+        }
+
+        // First check if the user provided a package version.
+        if (internalMetadataProvider.TryGetMetadata(MetadataKey.PackageSupplier, out object packageSupplier))
+        {
+            return packageSupplier as string;
+        }
+
+        // Right now we don't have any better way to version the package. Throw an exception for the user to
+        // provide a package version.
+        throw new ArgumentException($"Unable to generate a package supplier based on provided parameters. " +
+                                    $"Please provide the package supplier in the 'PackageSupplier' parameter.");
+    }
+
+    public static string GetDocumentNamespace(this IInternalMetadataProvider internalMetadataProvider)
+    {
+        if (internalMetadataProvider is null)
+        {
+            throw new ArgumentNullException(nameof(internalMetadataProvider));
+        }
+
+        return internalMetadataProvider.GetSBOMNamespaceUri();
+    }
+
+    public static string GetSwidTagId(this IInternalMetadataProvider internalMetadataProvider)
+    {
+        if (internalMetadataProvider is null)
+        {
+            throw new ArgumentNullException(nameof(internalMetadataProvider));
+        }
+
+        var rootPackageVersion = Uri.EscapeDataString(internalMetadataProvider.GetPackageVersion());
+        var packageSupplierFromMetadata = Uri.EscapeDataString(internalMetadataProvider.GetPackageSupplier());
+        var rootPackageName = Uri.EscapeDataString(internalMetadataProvider.GetPackageName());
+
+        var namespaceUri = new Uri(internalMetadataProvider.GetSBOMNamespaceUri());
+
+        // Generate a guid for the new swid tag Id.
+        var tagId = Guid.NewGuid().ToString();
+
+        var swidPurl = $"pkg:swid/{packageSupplierFromMetadata}/{namespaceUri.Host}/{rootPackageName}@{rootPackageVersion}?tag_id={tagId}";
+
+        return swidPurl;
+    }
+
+    public static string GetGenerationTimestamp(this IInternalMetadataProvider internalMetadataProvider)
+    {
+        if (internalMetadataProvider is null)
+        {
+            throw new ArgumentNullException(nameof(internalMetadataProvider));
+        }
+
+        if (internalMetadataProvider.TryGetMetadata(MetadataKey.GenerationTimestamp, out object generationTimestamp))
+        {
+            return generationTimestamp as string;
```

Comment:
What's the type of `generationTimestamp`? If it's always a string, then we can strongly type it in the call to `TryGetMetadata`. If it's not a string, then the cast will fail and we'll return `null`. Is that ok?